### PR TITLE
Trying to enforce no DML constraint in select Where and Orderby Clauses

### PR DESCRIPTION
**Pull request by zchenb**
Fixes https://github.com/edgedb/edgedb/issues/4952

### CODE REVIEW

File: edb/edgeql/compiler/stmt.py [hunk 1]
```diff
@@ -133,10 +133,14 @@ def compile_SelectQuery(
             forward_rptr=forward_rptr,
             ctx=sctx)
 
-        stmt.where = clauses.compile_where_clause(expr.where, ctx=sctx)
+        with sctx.new(context.ContextSwitchMode.NEW) as sctx_no_dml:
+            sctx_no_dml.disallow_dml = True
```
Comment (msullivan) [inline edb/edgeql/compiler/stmt.py:137]: I'd do this in `clauses` itself, for two reasons:
 * A new context is already getting made there, so it'll be a bit faster
 * But more importantly, we want this to be true for where clauses on updates and deletes also

Comment (msullivan) [reply edb/edgeql/compiler/stmt.py:137]: And I think limit/offset should get checked too

File: edb/edgeql/compiler/context.py [hunk 2]
```diff
@@ -537,6 +537,10 @@ class ContextLevel(compiler.ContextLevel):
     active_computeds: ordered.OrderedSet[s_pointers.Pointer]
     """A ordered set of currently compiling computeds"""
 
+    disallow_dml: bool
```
Comment (msullivan) [inline edb/edgeql/compiler/context.py:540]: Maybe make this an `Optional[str]` so we can say "UPDATE may not be done in a FILTER clause" or whatever

**Review by msullivan — COMMENTED** [review]
Good start!

File: edb/edgeql/compiler/stmt.py (hunk 1 above)
Comment (zchenb) [reply edb/edgeql/compiler/stmt.py:137]: I think that DMLs in limits and offsets are fine as they only get executed once.

**Review by msullivan — APPROVED** [review]
Great!

**Pull request merged**